### ISSUE TITLE
Build conditionally

### ISSUE THREAD

**Issue by idkq** [rrousselGit/provider#608]
I need the build method of a widget (screen) to be called only when a condition is satisfied, as opposed to whenever I access a variable.

I tried `.select()` but it does not work, since it does access the variable.

I tried to read the variable with `.read()` and if true, `.watch()` so it would reubuild, but `.read()` is not allowed in the build method.

Use case: I have a stack of screens and I only want to rebuild the screen if its visible, but it's being rebuilt many times in the background because I have a variable that holds the index of the visible screen.

**Comment by rrousselGit** [comment]
Do you have some code sample so I can better understand the issue?

**Comment by idkq** [comment]
Sure, here we go

```
class MyScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    
    // WRONG
    //forces rebuild if coming popping from another screen
    //ensuring accurate calculations from likely database changes
    context.select((MyAppState appState) {
      appState.activeScreen==MyScreenEnum.main;
      return false;
    });

    // WRONG
    //if (context.read<MyAppState>().activeScreen == MyScreenEnum.main);
    //  context.watch<MyAppState>().activeScreen;
...
```

**Comment by idkq** [comment]
In other words if the screen becomes active I want to rebuild (build method to be called). `appState.activeScreen` holds which screen is visible. If current screen is the active screen, the rebuild.

**Comment by idkq** [comment]
I am not sure if this has anything to do with Provider, but something related to Pages stacks that you actually helped solving (links below). I removed provider calls from screen/widget build and problem remains because I'm calling it inside a [RouterDelegate]. I will close this for now, but still investigating.

https://github.com/flutter/flutter/issues/11655 
https://github.com/flutter/flutter/issues/45797

**Comment by idkq** [comment]
Right. So my findings:

- `watch` cause all screen/pages stack to rebuild every time we navigate to A-> B-> C, which is unintended
- `read` not possible to use (as condition to determine which screen is visible) either inside the build or the callback from `select` so I can't really read without causing the whole stack to rebuild.

I think it is a 'bug' if we consider the new opaque rebuild condition that was implemented on the routes (links above ˆˆ).

Basically I can only rebuild all screens on stack, or none.

**Comment by rrousselGit** [comment]
Make sure you use const constructors on your pages:

```dart
MaterialApp(
  routes: {
    '/foo': (context) => const FooPage(),
  }
)
```

**Comment by idkq** [comment]
Not working, here is a repro https://gist.github.com/idkq/6ad5753b1562f78fe32b8c83cfe19b35

**Comment by rrousselGit** [comment]
Your gist doesn't use const constructors as I suggested

AppShell => const AppShell

**Comment by idkq** [comment]
Sorry I used all over but missed that one. Updated with const. https://gist.github.com/idkq/6ad5753b1562f78fe32b8c83cfe19b35 . Still does not work, it gets rebuilt every single time or alternatively if I comment line 300 `context.watch<BooksAppState>().selectedScreenIndex;` it will never be rebuilt. How does const help? Is there any way to unsubscribe the listener if the screen is not active?

**Comment by idkq** [comment]
`Provider.of<BooksAppState>(context, listen: myswitch).selectedScreenIndex;` doesn't work either. For some reason once it subscribes, even if listen: myswitch is false, it still listens (and rebuilds).

**Comment by rrousselGit** [comment]
I don't quite get what you're expecting here

The variable changes, so all the widgets listening to the variable rebuild. That sounds logical to me. What would you like?

**Comment by idkq** [comment]
Rebuild only if the variable changes to a specific value (i.e. selectedScreenIndex is equal active screen).

**Comment by rrousselGit** [comment]
And why is `select` not doing that?

**Comment by idkq** [comment]
Great thanks it worked, I implemented `select` in the wrong way as per my first message. I thought it would only listen to a part of a real instantiated object, in this case a primitive (built-in) int variable can't be divided into parts. But I understand now that "if they return a value different from the previously returned value, the dependent will be marked as needing to rebuild." (from docs)

I strongly suggest adding this to the docs as a use case since defining `select` as a method "which allows a widget to listen to only a small part of T" (from readme) is highly misleading IMHO.

Many thanks!

**Comment by idkq** [comment]
This is super hacky... I wish there was a better way. I had to create a second variable, which is an `array` of `int`s to hold the state of entering each screen `model.previousStateValue`. The solution you proposed returning a boolean state won't work because it will cause a rebuild when entering AND leaving each screen because the bool will switch on/off and will cause a rebuild each time for at least two screens.

Instead, the `int` will only be incremented if its entering the screen (not leaving).

Also, when building for the first time, it causes a rebuild twice and there is no way to prevent that it seems, because the `select` starts with no initial state apparently. Even returning 0.

```
    final isActiveScreen = context.select< BooksAppState, int>((model) {
      if (model.selectedScreenIndex == 0) {
        model.previousStateValue[0]++;
        // tries to prevent initial double building - doesnt work
        //if (model.previousStateValue[0]==1 || model.previousStateValue[0]==2)
        //  return 0;
      }
      return model.previousStateValue[0];
    });
```

**Comment by rrousselGit** [comment]
Rebuilding when leaving the screen sounds fine to me.

You're probably thinking too much.
